sif/internal/sif_util.go

```diff
@@ -0,0 +1,238 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/v5/sif/sif"
+	"github.com/pkg/errors"
+
+	imgspecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type SifImage struct {
+	fimg      sif.FileImage
+	rootfs    *sif.Descriptor
+	deffile   *sif.Descriptor
+	defReader *io.SectionReader
+	cmdlist   []string
+	runscript *bytes.Buffer
+	env       *sif.Descriptor
+	envReader *io.SectionReader
+	envlist   []string
+}
+
+func LoadSIFImage(path string) (image SifImage, err error) {
+	// open up the SIF file and get its header
+	image.fimg, err = sif.LoadContainer(path, true)
+	if err != nil {
+		return
+	}
+
+	// check for a system partition and save it
+	image.rootfs, _, err = image.fimg.GetPartPrimSys()
+	if err != nil {
+		return SifImage{}, errors.Wrap(err, "looking up rootfs from SIF file")
+	}
+
+	// look for a definition file object
+	searchDesc := sif.Descriptor{Datatype: sif.DataDeffile}
+	resultDescs, _, err := image.fimg.GetFromDescr(searchDesc)
+	if err == nil && resultDescs != nil {
+		// we assume in practice that typical SIF files don't hold multiple deffiles
+		image.deffile = resultDescs[0]
+		image.defReader = io.NewSectionReader(image.fimg.Fp, image.deffile.Fileoff, image.deffile.Filelen)
+	}
+	if err = image.generateConfig(); err != nil {
+		return SifImage{}, err
+	}
+
+	// look for an environment variable set object
+	searchDesc = sif.Descriptor{Datatype: sif.DataEnvVar}
+	resultDescs, _, err = image.fimg.GetFromDescr(searchDesc)
+	if err == nil && resultDescs != nil {
+		// we assume in practice that typical SIF files don't hold multiple EnvVar sets
+		image.env = resultDescs[0]
+		image.envReader = io.NewSectionReader(image.fimg.Fp, image.env.Fileoff, image.env.Filelen)
+	}
+
+	return image, nil
+}
+
+func (image *SifImage) parseEnvironment(scanner *bufio.Scanner) error {
+	for scanner.Scan() {
+		s := strings.TrimSpace(scanner.Text())
+		if s == "" || strings.HasPrefix(s, "#") {
+			continue
+		}
+		if strings.HasPrefix(s, "%") {
+			return nil
+		}
+		image.envlist = append(image.envlist, s)
+	}
+	if err := scanner.Err(); err != nil {
+		return errors.Wrap(err, "parsing environment from SIF definition file object")
+	}
+	return nil
+}
+
+func (image *SifImage) parseRunscript(scanner *bufio.Scanner) error {
+	for scanner.Scan() {
+		s := strings.TrimSpace(scanner.Text())
+		if strings.HasPrefix(s, "%") {
+			return nil
+		}
+		image.cmdlist = append(image.cmdlist, s)
+	}
+	if err := scanner.Err(); err != nil {
+		return errors.Wrap(err, "parsing runscript from SIF definition file object")
+	}
+	return nil
+}
+
+func (image *SifImage) generateRunscript() error {
+	base := `#!/bin/bash
+`
+	image.runscript = bytes.NewBufferString(base)
+	for _, s := range image.envlist {
+		_, err := image.runscript.WriteString(fmt.Sprintln(s))
+		if err != nil {
+			return errors.Wrap(err, "writing to runscript buffer")
+		}
+	}
+	for _, s := range image.cmdlist {
+		_, err := image.runscript.WriteString(fmt.Sprintln(s))
```

Comment:
⚠️ Is it intentional that `Sprintln` interprets Go format strings in the data? That seems very surprising.

Reply:
how so?

Reply:
What if cmdlist contains `Hello %s%50d%n`? That’s not going to result in anything _useful_ in the generated script AFAICS.

Luckily, with Go, it’s not a trivially-exploited arbitrary-code-execution vulnerability, but it’s both useless and it exposes the format string processing code to arbitrary potentially-malicious input, something it’s not really intended for.

Reply:
After two months I need to confess that I need to sit down and get back into the context, and I will do that sometime soon. Thanks for the thorough review BTW, but are we not talking about an Sprint(a) vs an Sprintf(format, a) construct here, where Sprintln(a) does not interpret anything? Or are you trying to make another point?

Reply:
You’re right, my mistake; this is not interpreting as a format string, so it is safe.

---

(It doesn’t quite make sense to me to do it this way — either performance and memory use is critical, in which case `runscript.WriteString(entry); runscript.WriteByte('\n')` (avoiding an allocated copy for the string with newline, and all the reflection implied by `Sprintln`)), or it isn’t so much, and those loops can be replaced by something like `strings.Join(image.cmdlist, "\n")`.

Assuming the performance of this not _that_ much of a concern, that can be deferred for some indeterminate future.)